src/Stratis.Bitcoin/Utilities/AsyncQueue.cs

```diff
@@ -0,0 +1,104 @@
+﻿using System;
+using System.Collections.Concurrent;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Stratis.Bitcoin.Utilities
+{
+    /// <summary>
+    /// Async queue is a simple thread-safe queue that asynchronously executes 
+    /// a user-defined callback when a new item is added to the queue.
+    /// <para>
+    /// The queue guarantees that the user-defined callback is executed only once at the time. 
+    /// If another item is added to the queue, the callback is called again after the current execution 
+    /// is finished.
+    /// </para>
+    /// </summary>
+    /// <typeparam name="T">Type of items to be inserted in the queue.</typeparam>
+    public class AsyncQueue<T>: IDisposable
+    {
+        /// <summary>
+        /// Represents a callback method to be executed when a new item is added to the queue.
+        /// </summary>
+        /// <param name="item">Newly added item.</param>
+        /// <param name="cancellationToken">Cancellation token that the callback method should use for its async operations to avoid blocking the queue during shutdown.</param>
+        public delegate Task OnEnqueueAsync(T item, CancellationToken cancellationToken);
+
+        /// <summary>Storage of items in the queue that are waiting to be consumed.</summary>
+        private readonly ConcurrentQueue<T> items;
```

Comment:
Can you please make it `protected`?

Reply:
if needed, the further work can change this

just as there is no need for manual trigger now, it is perfectly fine for this to be private

there is no problem changing anything in the class later and extend it to either that version with manual trigger that i have in separate branch, or to whatever you have in your mind. this being private does not forbid any changes later

Reply:
Ok, reasonable.